libflux/src/semantic/walk.rs

```diff
@@ -0,0 +1,1131 @@
+                walk(v, Node::from_expr(&mut n.left));
+                walk(v, Node::from_expr(&mut n.right));
+            }
+            Node::UnaryExpr(ref mut n) => {
+                walk(v, Node::from_expr(&mut n.argument));
+            }
+            Node::CallExpr(ref mut n) => {
+                walk(v, Node::from_expr(&mut n.callee));
+                if let Some(mut p) = n.pipe.as_mut() {
+                    walk(v, Node::from_expr(&mut p));
+                }
+                for mut arg in n.arguments.iter_mut() {
+                    walk(v, Node::Property(&mut arg));
+                }
+            }
+            Node::ConditionalExpr(ref mut n) => {
+                walk(v, Node::from_expr(&mut n.test));
+                walk(v, Node::from_expr(&mut n.consequent));
+                walk(v, Node::from_expr(&mut n.alternate));
+            }
+            Node::StringExpr(ref mut n) => {
+                for mut part in n.parts.iter_mut() {
+                    walk(v, Node::from_string_expr_part(&mut part));
+                }
+            }
+            Node::IntegerLit(_) => {}
+            Node::FloatLit(_) => {}
+            Node::StringLit(_) => {}
+            Node::DurationLit(_) => {}
+            Node::UintLit(_) => {}
+            Node::BooleanLit(_) => {}
+            Node::DateTimeLit(_) => {}
+            Node::RegexpLit(_) => {}
+            Node::ExprStmt(ref mut n) => {
+                walk(v, Node::from_expr(&mut n.expression));
+            }
+            Node::OptionStmt(ref mut n) => {
+                walk(v, Node::from_assignment(&mut n.assignment));
+            }
+            Node::ReturnStmt(ref mut n) => {
+                walk(v, Node::from_expr(&mut n.argument));
+            }
+            Node::TestStmt(ref mut n) => {
+                walk(v, Node::VariableAssgn(&mut n.assignment));
+            }
+            Node::BuiltinStmt(ref mut n) => {
+                walk(v, Node::Identifier(&mut n.id));
+            }
+            Node::Block(ref mut n) => {
+                for mut s in n.body.iter_mut() {
+                    walk(v, Node::from_stmt(&mut s));
+                }
+            }
+            Node::Property(ref mut n) => {
+                walk(v, Node::Identifier(&mut n.key));
+                walk(v, Node::from_expr(&mut n.value));
+            }
+            Node::TextPart(_) => {}
+            Node::InterpolatedPart(ref mut n) => {
+                walk(v, Node::from_expr(&mut n.expression));
+            }
+            Node::VariableAssgn(ref mut n) => {
+                walk(v, Node::Identifier(&mut n.id));
+                walk(v, Node::from_expr(&mut n.init));
+            }
+            Node::MemberAssgn(ref mut n) => {
+                walk(v, Node::MemberExpr(&mut n.member));
+                walk(v, Node::from_expr(&mut n.init));
+            }
+        };
+    }
+    v.done(node);
+}
+
+impl<F> Visitor for F
+where
+    F: for<'a> FnMut(Node<'a>),
```

Comment:
@jlapacik you'll love this 😄

Reply:
@nathanielc this is Higher-Ranked Trait Bounds (woah).

According to the book: 

> This lets the Rust compiler find the minimum lifetime to invoke our closure and satisfy the borrow checker's rules.

The problem here happens when you use that closure.

```
walk<'a>(&mut v: FnMut(Node<'a>), node: Node<???IDK???>) {
  ...
}
```

> When a function has an explicit lifetime parameter, that lifetime must be at least as long as the _entire_ call to that function.

It seems that, due to lifetime elision, the compiler cannot determine the lifetime of `node` compared to the parameter of the visitor and it cannot know that it lives more than the function.

```
walk(&mut v: for <'a> FnMut(Node<'a>), node: Node<???IDK???>) {
  ...
}
```

This is ok, because `<'a>` is not associated to the lifetime of the function call, but only to the node itself ---> the compiler can find a good lifetime to go on.


That's rather obscure to me now. I plan to study more lifetimes and try different combinations to see what works. However, this solution seems correct.

Reply:
Thanks, maybe a quick comment with a link to https://doc.rust-lang.org/nomicon/hrtb.html will help future readers learn why this is the case for closures? When I first read the code I didn't think it was specific to closures and so it was harder to understand its meaning. (I still don't fully understand, but I see why closures need special handling).

Reply:
Can you explain this a little more. I don't see why the lifetime of Node is less than that of the visitor?